File: app/src/application/EnnemiDAO.java
package application;

import objets.*;
import personnages.Boss;
import personnages.Ennemi;

import java.sql.*;
import java.util.ArrayList;

public class EnnemiDAO {
    private final Connection connection;
    private Donjon donjon;

    public EnnemiDAO(Connection connection) {
        this.connection = connection;
    }

    public ArrayList<Ennemi> recupererEnnemis(Donjon donjon) throws SQLException {
        ArrayList<Ennemi> ennemis = new ArrayList<>();
        this.donjon = donjon;

        // Étape 1 : Récupérer les informations de l'ennemi
        String selectEnnemiSQL = "SELECT * FROM ennemis WHERE donjon = ?";
        try (PreparedStatement pstmt = connection.prepareStatement(selectEnnemiSQL)) {
            pstmt.setInt(1, donjon.getNiveau());
            ResultSet rs = pstmt.executeQuery();

            while (rs.next()) {
                ennemis.add(new Ennemi(
                        rs.getString("nom"),
                        rs.getInt("ptsVie"),
                        rs.getInt("niveau"),
                        rs.getInt("attaque"),
                        rs.getDouble("poidsSpawn"),
                        rs.getInt("xpDrop"),
                        recupererDropsPourEnnemi(rs.getInt("id"), "ennemi")));
            }
        }
        return ennemis;
    }

    private ArrayList<Objet> recupererDropsPourEnnemi(int ennemiId, String typeEnnemi) throws SQLException {
        ArrayList<Objet> drops = new ArrayList<>();
        String selectDropTypesSQL = "SELECT drop_id, drop_type FROM " + typeEnnemi + "_drops WHERE " + typeEnnemi + "_id = ?";

        try (PreparedStatement pstmt = connection.prepareStatement(selectDropTypesSQL)) {
            pstmt.setInt(1, ennemiId);
            ResultSet rs = pstmt.executeQuery();

            while (rs.next()) {
                int dropId = rs.getInt("drop_id");
                String dropType = rs.getString("drop_type");

                switch (dropType) {
                    case "arme" -> {
                        // Préparer le statement pour récupérer les informations d'arme
                        String selectArmeSQL = "SELECT nom, description, drop_rate, degats, effetStatut FROM armes WHERE id = ?";
                        try (PreparedStatement pstmtArme = connection.prepareStatement(selectArmeSQL)) {
                            pstmtArme.setInt(1, dropId);
                            ResultSet rsArme = pstmtArme.executeQuery();

                            if (rsArme.next()) {
                                drops.add(new Arme(
                                        rsArme.getString("nom"),
                                        rsArme.getString("description"),
                                        rsArme.getDouble("drop_rate"),
                                        rsArme.getInt("degats"),
                                        rsArme.getString("effetStatut")));
                            }
                        }
                    }
                    case "armure" -> {
                        // Préparer le statement pour récupérer les informations d'armure
                        String selectArmureSQL = "SELECT nom, description, drop_rate, ptsArmure FROM armures WHERE id = ?";
                        try (PreparedStatement pstmtArmure = connection.prepareStatement(selectArmureSQL)) {
                            pstmtArmure.setInt(1, dropId);
                            ResultSet rsArmure = pstmtArmure.executeQuery();

                            if (rsArmure.next()) {
                                drops.add(new Armure(
                                        rsArmure.getString("nom"),
                                        rsArmure.getString("description"),
                                        rsArmure.getInt("ptsArmure"),
                                        rsArmure.getDouble("drop_rate")));
                            }
                        }
                    }
                    case "potion" -> {
                        // Préparer le statement pour récupérer les informations de potion
                        String selectPotionSQL = "SELECT nom, soin, drop_rate FROM potions WHERE id = ?";
                        try (PreparedStatement pstmtPotion = connection.prepareStatement(selectPotionSQL)) {
                            pstmtPotion.setInt(1, dropId);
                            ResultSet rsPotion = pstmtPotion.executeQuery();

                            if (rsPotion.next()) {
                                drops.add(new Potion(
                                        rsPotion.getString("nom"),
                                        rsPotion.getInt("soin"),
                                        rsPotion.getDouble("drop_rate")));
                            }
                        }
                    }
                    case "divers" -> {
                        // Préparer le statement pour récupérer les informations de divers
                        String selectInvocationSQL = "SELECT nom, description, drop_rate, boss_id FROM invocation_boss WHERE id = ?";
                        try (PreparedStatement pstmtInvocation = connection.prepareStatement(selectInvocationSQL)) {
                            pstmtInvocation.setInt(1, dropId);
                            ResultSet rsInvocation = pstmtInvocation.executeQuery();

                            if (rsInvocation.next()) {
                                drops.add(new ObjetInvoqueBoss(
                                        rsInvocation.getString("nom"),
                                        rsInvocation.getString("description"),
                                        rsInvocation.getDouble("drop_rate"),
                                        recupererBoss(rsInvocation.getInt("boss_id")),
                                        donjon));
                            }
                        }
                    }
                }
            }
        }
        return Objet.pondererDropRates(drops);
    }

    private Boss recupererBoss(int bossKey) throws SQLException
    {
        String selectBossSQL = "SELECT * FROM bosses WHERE id = ?";
        try (PreparedStatement pstmtInvocation = connection.prepareStatement(selectBossSQL)) {
            pstmtInvocation.setInt(1, bossKey);
            ResultSet rsBoss = pstmtInvocation.executeQuery();
            if (rsBoss.next())
            {
                return new Boss(
                        rsBoss.getString("nom"),
                        rsBoss.getString("description"),
                        rsBoss.getInt("donjon"),
                        rsBoss.getInt("ptsVie"),
                        rsBoss.getInt("attaque"),
                        rsBoss.getInt("xp_drop"),
                        rsBoss.getInt("ptsArmure"),
                        recupererDropsPourEnnemi(bossKey, "boss"));
            }
        }

        return null;
    }
}

File: KF-Remake/Kingdom-Fall/source/application/EnnemiDAO.cs
using System.Collections;
using personnages;
using objets;
using MySql.Data.MySqlClient;
namespace App;

class EnnemiDAO {
    private Donjon donjon;
    private DatabaseManager dbm;


    public EnnemiDAO(DatabaseManager dbm, Donjon donjon)
    {
        this.dbm = dbm;
        this.donjon = donjon;
    }

    public List<Ennemi> RecupererEnnemis()
    {
        List<Ennemi> ennemis = new List<Ennemi>();
        // Étape 1 : Récupérer les informations de l'ennemi
        string selectEnnemiSQL = $"SELECT * FROM ennemis WHERE donjon = {donjon.niveau}";
        
        dbm.OpenConnection();

        using (MySqlDataReader reader = dbm.ExecuteQuery(selectEnnemiSQL))
        {
            while (reader.Read())
            {
                ennemis.Add(new Ennemi(
                            reader.GetString("nom"),
                            reader.GetInt32("ptsVie"),
                            reader.GetInt32("niveau"),
                            reader.GetInt32("attaque"),
                            reader.GetDouble("poidsSpawn"),
                            reader.GetInt32("xpDrop"),
                            RecupererDropsPourEnnemi(reader.GetInt32("id"), "ennemi", reader.GetString("nom"))));
            }
        }
        

        dbm.CloseConnection();

        return ennemis;
    }


    private List<Objet> RecupererDropsPourEnnemi(int ennemiId, string classeEnnemi, string typeEnnemi)
    {
        List<Objet> drops = new List<Objet>();
        string selectDropTypesSQL = $"SELECT drop_id, drop_type FROM {classeEnnemi}_drops WHERE {classeEnnemi}_id = {ennemiId}";

        dbm.OpenConnection();
        using (var reader = dbm.ExecuteQuery(selectDropTypesSQL))
        {
            while (reader.Read())
            {
                int dropId = reader.GetInt32("drop_id");
                string dropType = reader.GetString("drop_type");
                dbm.OpenConnection();

                switch (dropType)
                {
                    case "arme":
                        string selectArmeSQL = $"SELECT nom, description, drop_rate, degats, effetStatut FROM armes WHERE id = '{dropId}'";
                        using (var armeReader = dbm.ExecuteQuery(selectArmeSQL))
                        {
                            if (armeReader.Read())
                            {
                                drops.Add(new Arme(
                                            armeReader.GetString("nom"),
                                            armeReader.GetString("description"),
                                            armeReader.GetDouble("drop_rate"),
                                            armeReader.GetInt32("degats"),
                                            armeReader.GetString("effetStatut")));
                            }
                        }
                        break;

                    case "armure":
                        string selectArmureSQL = $"SELECT nom, description, drop_rate, ptsArmure FROM armures WHERE id = {dropId}";
                        using (var armureReader = dbm.ExecuteQuery(selectArmureSQL))
                        {
                            if (armureReader.Read())
                            {
                                drops.Add(new Armure(
                                            armureReader.GetString("nom"),
                                            armureReader.GetString("description"),
                                            armureReader.GetInt32("ptsArmure"),
                                            armureReader.GetDouble("drop_rate")));
                            }
                        }
                        break;

                    case "potion":
                        string selectPotionSQL = $"SELECT nom, soin, drop_rate FROM potions WHERE id = {dropId}";

                        using (var potionReader = dbm.ExecuteQuery(selectPotionSQL))
                        {
                            if (potionReader.Read())
                            {
                                drops.Add(new Potion(
                                            potionReader.GetString("nom"),
                                            potionReader.GetInt32("soin"),
                                            potionReader.GetDouble("drop_rate")));
                            }
                        }
                        break;

                    case "divers":
                        string selectInvocationSQL = $"SELECT nom, description, drop_rate, boss_id FROM invocation_boss WHERE id = {dropId}";
                        using (var invocationReader = dbm.ExecuteQuery(selectInvocationSQL))
                        {
                            if (invocationReader.Read())
                            {
                                drops.Add(new ObjetInvoqueBoss(
                                            invocationReader.GetString("nom"),
                                            invocationReader.GetString("description"),
                                            invocationReader.GetDouble("drop_rate"),
                                            RecupererBoss(invocationReader.GetInt32("boss_id")),
                                            donjon));

                                dbm.CloseConnection();
                            }
                        }
                        break;
                }
                dbm.CloseConnection();
            }
        
        }
        dbm.CloseConnection();
        return Objet.PondererDropRates(drops, typeEnnemi);
    }

    private Boss RecupererBoss(int bossKey)
    {
        string selectBossSQL = $"SELECT * FROM bosses WHERE id = {bossKey}";
        dbm.OpenConnection();


        using var reader = dbm.ExecuteQuery(selectBossSQL);
        if (reader.Read())
        {
            return new Boss(
                    reader.GetString("nom"),
                    reader.GetString("description"),
                    reader.GetInt32("donjon"),
                    reader.GetInt32("ptsVie"),
                    reader.GetInt32("attaque"),
                    reader.GetInt32("xp_drop"),
                    reader.GetInt32("ptsArmure"),
                    RecupererDropsPourEnnemi(bossKey, "boss", reader.GetString("nom")));
        }

        dbm.CloseConnection();
        throw new KeyNotFoundException($"Boss avec ID {bossKey} non trouvé dans la base de données.");


    }
}
